pkg/registration/hub/clusterprofile/controller.go

```diff
@@ -0,0 +1,171 @@
+		clusterProfileClient: clusterProfileClient,
+		clusterProfileLister: clusterProfileInformer.Lister(),
+		patcher: patcher.NewPatcher[
+			*cpv1alpha1.ClusterProfile, cpv1alpha1.ClusterProfileSpec, cpv1alpha1.ClusterProfileStatus](
+			clusterProfileClient.ApisV1alpha1().ClusterProfiles(ClusterProfileNamespace)),
+		eventRecorder: recorder.WithComponentSuffix("cluster-profile-controller"),
+	}
+
+	return factory.New().
+		WithInformersQueueKeysFunc(queue.QueueKeyByMetaName, clusterInformer.Informer(), clusterProfileInformer.Informer()).
+		WithSync(c.sync).
+		ToController("ClusterProfileController", recorder)
+}
+
+func (c *clusterProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling Cluster", "ClusterName", managedClusterName)
+
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if !managedCluster.DeletionTimestamp.IsZero() {
+		// TODO: remove clusterprofile
+		return nil
+	}
+
+	clusterProfile, err := c.clusterProfileLister.ClusterProfiles(ClusterProfileNamespace).Get(managedClusterName)
+	// create cluster profile if not found
+	if errors.IsNotFound(err) {
+		clusterProfile = &cpv1alpha1.ClusterProfile{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   managedClusterName,
+				Labels: map[string]string{cpv1alpha1.LabelClusterManagerKey: ClusterProfileManagerName},
+			},
+			Spec: cpv1alpha1.ClusterProfileSpec{
+				DisplayName: managedClusterName,
+				ClusterManager: cpv1alpha1.ClusterManager{
+					Name: ClusterProfileManagerName,
+				},
+			},
+		}
+		_, err = c.clusterProfileClient.ApisV1alpha1().ClusterProfiles(ClusterProfileNamespace).Create(ctx, clusterProfile, metav1.CreateOptions{})
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	// return if not managed by ocm
+	if clusterProfile.Spec.ClusterManager.Name != ClusterProfileManagerName {
+		logger.Info("Not managed by open-cluster-management, skipping", "ClusterName", managedClusterName)
+		return nil
+	}
+
+	newClusterProfile := clusterProfile.DeepCopy()
+
+	// sync required labels
+	mclLabels := managedCluster.GetLabels()
+	mclSetLabel := mclLabels[v1beta2.ClusterSetLabel]
+	// The value of label "x-k8s.io/cluster-manager" MUST be the same as the name of the cluster manager.
+	requiredLabels := map[string]string{
+		cpv1alpha1.LabelClusterManagerKey: newClusterProfile.Spec.ClusterManager.Name,
+		cpv1alpha1.LabelClusterSetKey:     mclSetLabel,
+	}
+	modified := false
+	resourcemerge.MergeMap(&modified, &newClusterProfile.Labels, requiredLabels)
+
+	// patch labels
+	if modified {
+		_, err := c.patcher.PatchLabelAnnotations(ctx, newClusterProfile, newClusterProfile.ObjectMeta, clusterProfile.ObjectMeta)
+		if err != nil {
+			return err
```

Comment:
directly return here, we can update status in the next loop.

Reply:
fixed.